From 19ad509f7b61ef75bfb7a0b93f6246db1c78826d Mon Sep 17 00:00:00 2001
From: Jakob Kummerow <jkummerow@chromium.org>
Date: Thu, 06 Feb 2025 14:18:48 +0100
Subject: [PATCH] Merged: [wasm] Replace {dead_code_} set with {is_dying_} bit

This saves some memory, and fixes a bug.

Fixed: 391907159
(cherry picked from commit 33ca4f51e5dbba9817eba16fd3249e66a880cf33)

Change-Id: I3324f33bbf7b6a191b508e854ef92c9e9a5efba5
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/6239495
Reviewed-by: Matthias Liedtke <mliedtke@chromium.org>
Commit-Queue: Matthias Liedtke <mliedtke@chromium.org>
Auto-Submit: Jakob Kummerow <jkummerow@chromium.org>
Commit-Queue: Jakob Kummerow <jkummerow@chromium.org>
Cr-Commit-Position: refs/branch-heads/13.3@{#38}
Cr-Branched-From: 41dacffe436aeb9311879cb07648f1e36609a804-refs/heads/13.3.415@{#1}
Cr-Branched-From: 3348638c0af67c885b30891a358c89a917ac9759-refs/heads/main@{#97937}
---

--- a/v8/src/wasm/wasm-code-manager.h
+++ b/v8/src/wasm/wasm-code-manager.h
@@ -321,6 +321,9 @@
     return ForDebuggingField::decode(flags_);
   }
 
+  bool is_dying() const { return dying_; }
+  void mark_as_dying() { dying_ = true; }
+
   // Returns {true} for Liftoff code that sets up a feedback vector slot in its
   // stack frame.
   // TODO(jkummerow): This can be dropped when we ship Wasm inlining.
@@ -450,6 +453,10 @@
   using ForDebuggingField = ExecutionTierField::Next<ForDebugging, 2>;
   using FrameHasFeedbackSlotField = ForDebuggingField::Next<bool, 1>;
 
+  // Will be set to {true} the first time this code object is considered
+  // "potentially dead" (to be confirmed by the next Wasm Code GC cycle).
+  std::atomic<bool> dying_{false};
+
   // WasmCode is ref counted. Counters are held by:
   //   1) The jump table / code table.
   //   2) {WasmCodeRefScope}s.
--- a/v8/src/wasm/wasm-engine.cc
+++ b/v8/src/wasm/wasm-engine.cc
@@ -1738,7 +1738,6 @@
   }
   // If any code objects are currently tracked as dead or near-dead, remove
   // references belonging to the NativeModule that's being deleted.
-  std::erase_if(dead_code_, part_of_native_module);
   std::erase_if(potentially_dead_code_, part_of_native_module);
 
   if (native_module->log_code()) DisableCodeLogging(native_module);
@@ -1824,9 +1823,11 @@
 
 bool WasmEngine::AddPotentiallyDeadCode(WasmCode* code) {
   base::SpinningMutexGuard guard(&mutex_);
-  if (dead_code_.contains(code)) return false;  // Code is already dead.
+  if (code->is_dying()) return false;
   auto added = potentially_dead_code_.insert(code);
-  if (!added.second) return false;  // An entry already existed.
+  DCHECK(added.second);
+  USE(added);
+  code->mark_as_dying();
   new_potentially_dead_code_size_ += code->instructions().size();
   if (v8_flags.wasm_code_gc) {
     // Trigger a GC if 64kB plus 10% of committed code are potentially dead.
@@ -1872,19 +1873,17 @@
     const std::vector<WasmCode*>& code_vec = dead_code_entry.second;
     TRACE_CODE_GC("Freeing %zu code object%s of module %p.\n", code_vec.size(),
                   code_vec.size() == 1 ? "" : "s", native_module);
-    for (WasmCode* code : code_vec) {
-      DCHECK(dead_code_.contains(code));
-      dead_code_.erase(code);
-    }
+#if DEBUG
+    for (WasmCode* code : code_vec) DCHECK(code->is_dying());
+#endif  // DEBUG
     native_module->FreeCode(base::VectorOf(code_vec));
   }
   if (dead_wrappers.size()) {
     TRACE_CODE_GC("Freeing %zu wrapper%s.\n", dead_wrappers.size(),
                   dead_wrappers.size() == 1 ? "" : "s");
-    for (WasmCode* code : dead_wrappers) {
-      DCHECK(dead_code_.contains(code));
-      dead_code_.erase(code);
-    }
+#if DEBUG
+    for (WasmCode* code : dead_wrappers) DCHECK(code->is_dying());
+#endif  // DEBUG
     GetWasmImportWrapperCache()->Free(dead_wrappers);
   }
 }
@@ -1974,16 +1973,15 @@
   if (!current_gc_info_->outstanding_isolates.empty()) return;
 
   // All remaining code in {current_gc_info->dead_code} is really dead.
-  // Move it from the set of potentially dead code to the set of dead code,
-  // and decrement its ref count.
+  // Remove it from the set of potentially dead code, and decrement its
+  // ref count.
   size_t num_freed = 0;
   DeadCodeMap dead_code;
   std::vector<WasmCode*> dead_wrappers;
   for (WasmCode* code : current_gc_info_->dead_code) {
     DCHECK(potentially_dead_code_.contains(code));
+    DCHECK(code->is_dying());
     potentially_dead_code_.erase(code);
-    DCHECK(!dead_code_.contains(code));
-    dead_code_.insert(code);
     if (code->DecRefOnDeadCode()) {
       NativeModule* native_module = code->native_module();
       if (native_module) {
@@ -2007,7 +2005,7 @@
 }
 
 size_t WasmEngine::EstimateCurrentMemoryConsumption() const {
-  UPDATE_WHEN_CLASS_CHANGES(WasmEngine, 728);
+  UPDATE_WHEN_CLASS_CHANGES(WasmEngine, 688);
   UPDATE_WHEN_CLASS_CHANGES(IsolateInfo, 168);
   UPDATE_WHEN_CLASS_CHANGES(NativeModuleInfo, 56);
   UPDATE_WHEN_CLASS_CHANGES(CurrentGCInfo, 96);
@@ -2018,7 +2016,6 @@
     result += ContentSize(async_compile_jobs_);
     result += async_compile_jobs_.size() * sizeof(AsyncCompileJob);
     result += ContentSize(potentially_dead_code_);
-    result += ContentSize(dead_code_);
 
     // TODO(14106): Do we care about {compilation_stats_}?
     // TODO(14106): Do we care about {code_tracer_}?
--- a/v8/src/wasm/wasm-engine.h
+++ b/v8/src/wasm/wasm-engine.h
@@ -489,12 +489,8 @@
   size_t new_potentially_dead_code_size_ = 0;
   // Set of potentially dead code. This set holds one ref for each code object,
   // until code is detected to be really dead. At that point, the ref count is
-  // decremented and code is moved to the {dead_code} set. If the code is
-  // finally deleted, it is also removed from {dead_code}.
+  // decremented and code is removed from the set.
   std::unordered_set<WasmCode*> potentially_dead_code_;
-  // Code that is not being executed in any isolate any more, but the ref count
-  // did not drop to zero yet.
-  std::unordered_set<WasmCode*> dead_code_;
   int8_t num_code_gcs_triggered_ = 0;
 
   // If an engine-wide GC is currently running, this pointer stores information
--- a/v8/src/wasm/wasm-import-wrapper-cache.cc
+++ b/v8/src/wasm/wasm-import-wrapper-cache.cc
@@ -145,6 +145,8 @@
       GetProcessWideWasmCodePointerTable()->GetEntrypointWithoutSignatureCheck(
           call_target));
   if (iter == codes_.end()) return nullptr;
+  WasmCodeRefScope::AddRef(iter->second);
+  if (iter->second->is_dying()) return nullptr;
   return iter->second;
 }
 
@@ -161,7 +163,10 @@
     // Now that we have the lock (in the form of the cache_scope), check
     // again whether another thread has just created the wrapper.
     wasm_code = cache_scope[key];
-    if (wasm_code) return wasm_code;
+    if (wasm_code) {
+      WasmCodeRefScope::AddRef(wasm_code);
+      if (!wasm_code->is_dying()) return wasm_code;
+    }
 
     wasm_code = cache_scope.AddWrapper(key, std::move(result),
                                        WasmCode::Kind::kWasmToJsWrapper);
@@ -225,6 +230,7 @@
   auto it = entry_map_.find({kind, type_index, expected_arity, suspend});
   if (it == entry_map_.end()) return nullptr;
   WasmCodeRefScope::AddRef(it->second);
+  if (it->second->is_dying()) return nullptr;
   return it->second;
 }
 
@@ -239,6 +245,9 @@
   DCHECK_EQ(candidate->instruction_start(), iter->first);
   if (!candidate->contains(pc)) return nullptr;
   WasmCodeRefScope::AddRef(candidate);
+  // Note: this function is used for iterating the stack, where dying
+  // code objects can still have their last few activations, so we
+  // must return {candidate} even if {candidate->is_dying()}.
   return candidate;
 }
 
--- a/v8/src/wasm/wasm-objects.cc
+++ b/v8/src/wasm/wasm-objects.cc
@@ -568,6 +568,7 @@
   if (target_instance_data->dispatch_table_for_imports()->IsAWrapper(
           func->func_index)) {
     uint64_t signature_hash = wasm::SignatureHasher::Hash(func->sig);
+    wasm::WasmCodeRefScope code_ref_scope;
     dispatch_table->SetForWrapper(
         entry_index, *implicit_arg,
         wasm::GetProcessWideWasmCodePointerTable()->GetEntrypoint(
@@ -2219,7 +2220,7 @@
                                       wasm::WasmCode* compiled_wrapper,
                                       NewOrExistingEntry new_or_existing) {
   DCHECK_NE(implicit_arg, Smi::zero());
-
+  SBXCHECK(!compiled_wrapper || !compiled_wrapper->is_dying());
   SBXCHECK_BOUNDS(index, length());
   DCHECK(IsWasmImportData(implicit_arg) ||
          IsWasmTrustedInstanceData(implicit_arg));
@@ -2398,8 +2399,15 @@
       // After installing a compiled wrapper, we don't set a call origin
       // any more.
       if (import_data->has_call_origin()) {
-        DCHECK_EQ(*old_table, import_data->call_origin());
-        import_data->set_call_origin(*new_table);
+        if (import_data->call_origin() == *old_table) {
+          import_data->set_call_origin(*new_table);
+        } else {
+#if DEBUG
+          wasm::WasmCodeRefScope code_ref_scope;
+          DCHECK_NOT_NULL(
+              wasm::GetWasmImportWrapperCache()->FindWrapper(call_target));
+#endif  // DEBUG
+        }
       }
     }
 
--- a/v8/test/cctest/wasm/wasm-run-utils.cc
+++ b/v8/test/cctest/wasm/wasm-run-utils.cc
@@ -294,6 +294,7 @@
   }
 
   if (function_indexes) {
+    WasmCodeRefScope code_ref_scope;
     for (uint32_t i = 0; i < table_size; ++i) {
       uint32_t function_index = function_indexes[i];
       WasmFunction& function = test_module_->functions[function_index];
